libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -799,22 +799,33 @@ bool IRGeneratorForStatements::visit(BinaryOperation const& _binOp)
 
 	if (TokenTraits::isCompareOp(op))
 	{
-		if (auto type = dynamic_cast<FunctionType const*>(commonType))
-		{
-			solAssert(op == Token::Equal || op == Token::NotEqual, "Invalid function pointer comparison!");
-			solAssert(type->kind() != FunctionType::Kind::External, "External function comparison not allowed!");
-		}
-
 		solAssert(commonType->isValueType(), "");
+
 		bool isSigned = false;
 		if (auto type = dynamic_cast<IntegerType const*>(commonType))
 			isSigned = type->isSigned();
 
 		string args = expressionAsType(_binOp.leftExpression(), *commonType, true);
 		args += ", " + expressionAsType(_binOp.rightExpression(), *commonType, true);
 
+		auto functionType = dynamic_cast<FunctionType const*>(commonType);
+		solAssert(functionType ? (op == Token::Equal || op == Token::NotEqual) : true, "Invalid function pointer comparison!");
+
 		string expr;
-		if (op == Token::Equal)
+
+		if (functionType && functionType->kind() ==  FunctionType::Kind::External)
+		{
+			expr = m_utils.externalFunctionPointersEqualFunction() +
+				"(" +
+				IRVariable{_binOp.leftExpression()}.part("address").name() + "," +
+				IRVariable{_binOp.leftExpression()}.part("functionSelector").name() + "," +
+				IRVariable{_binOp.rightExpression()}.part("address").name() + "," +
+				IRVariable{_binOp.rightExpression()}.part("functionSelector").name() +
+				")";
```

Comment:
Looking at this once more, it may also have been nice to build this as
```suggestion
			expr = m_utils.externalFunctionPointersEqualFunction(functionType) +
				"(" +
				IRVariable{_binOp.leftExpression()}.commaSeparatedList() +
				"," +
				IRVariable{_binOp.rightExpression()}.commaSeparatedList() +
				")";
```
and then have ``externalFunctionPointersEqualFunction`` assert that the function type has only two stack slots...

But I'd say we can keep things like they are for now and change this, if we ever want to compare multi-slot function types (with additional gas slots, etc. - which I'm not sure we will, since it's probably not particularly useful).

So instead a simple assertion for the type having ``sizeOnStack() == 2`` will also be fine.